File: 543. Diameter of Binary Tree/Diameter of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 */
#include <algorithm>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int diameter=0;
    int maxH(TreeNode* node){
        if(!node) return 0;

        int L_H=maxH(node->left);
        int R_H=maxH(node->right);

        diameter=max(diameter,L_H+R_H);

        return max(L_H,R_H)+1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        maxH(root);
        return diameter;
    }
};

File: First 100 completed solutions/543. Diameter of Binary Tree/Diameter of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.diameter=0
        def maxH(node):
            if not node:
                return 0
            
            L_H=maxH(node.left)
            R_H=maxH(node.right)
            self.diameter = max(self.diameter, L_H + R_H)

            return max(L_H,R_H)+1
        maxH(root)
        return self.diameter

        